### ISSUE TITLE
False positives for RS1024 (Compare symbols correctly)

### ISSUE THREAD

**Issue by KrisVandermotten** [dotnet/roslyn-analyzers#4410]
Comparing instances of `ITypeSymbol` using their `IEquatable<ISymbol>` interface could be a bug, because of NRT. RS1024 was introduced to discover those possible bugs early, and force the author to be explicit about the symbol equality comparer to be used. Obviously, the rule should apply when comparing `ITypeSymbol` instances, but also when comparing `ISymbol` instances, since those just might be `ITypeSymbol` instances.

However, the rule should not apply to types implementing `IEquatable<ISymbol>` that do not inherit from `ITypeSymbol`. Examples include `IAssemblySymbol`, `IParameterSymbol` and others.

### Steps To Reproduce

Consider this code:

```cs
    private void M(SyntaxNodeAnalysisContext context, ISymbol symbol)
    {
        if (symbol.ContainingAssembly.Equals(context.Compilation.Assembly))
        {
            // ...
        }
    }
```

### Expected behavior

No diagnostic is reported, since there is no ambiguity in comparing `IAssemblySymbol` instances.

### Actual behavior

    warning RS1024: Compare symbols correctly

**Comment by sharwell** [comment]
@333fred @mavasani Do you know which symbols can never have nullability information?

**Comment by KrisVandermotten** [comment]
> Do you know which symbols can never have nullability information?

I would suggest to turn that question around: which symbols can have nullability information?
The best way to avoid false positives is to be explicit about the symbol types that can have diagnostics, and to ignore all other types.

It's clear that calling the `Equals<ISymbol>` method on expressions of the following types is problematic:
- ISymbol (and ISymbol?)
- ITypeSymbol (and ITypeSymbol?)
- INamedTypeSymbol and (INamedTypeSymbol?)

As far as I know, it does not apply to expressions of type `IPointerTypeSymbol`. I must admit I'm not entirely sure about `IMethodSymbol`. But for all other symbol types, I think RS1024 could never apply.

**Comment by sharwell** [comment]
It looks this applies to all member symbols (`IFieldSymbol`, `IEventSymbol`, `IMethodSymbol`, `IParameterSymbol`, etc.). This narrows down the potential false positives to:

* IAssemblySymbol
* IModuleSymbol
* INamespaceSymbol

**Comment by KrisVandermotten** [comment]
What about `IParameterSymbol` ?

**Comment by mavasani** [comment]
Tagging @Evangelink
